lang/src/org/partiql/lang/planner/transforms/LogicalToLogicalResolvedVisitorTransform.kt

```diff
@@ -0,0 +1,393 @@
+package org.partiql.lang.planner.transforms
+
+import com.amazon.ionelement.api.ionSymbol
+import org.partiql.lang.ast.sourceLocation
+import org.partiql.lang.domains.PartiqlLogical
+import org.partiql.lang.domains.PartiqlLogicalResolved
+import org.partiql.lang.domains.PartiqlLogicalToPartiqlLogicalResolvedVisitorTransform
+import org.partiql.lang.domains.toBindingCase
+import org.partiql.lang.errors.Problem
+import org.partiql.lang.errors.ProblemHandler
+import org.partiql.lang.eval.BindingName
+import org.partiql.lang.eval.builtins.DYNAMIC_LOOKUP_FUNCTION_NAME
+import org.partiql.lang.planner.GlobalBindings
+import org.partiql.lang.planner.ResolutionResult
+import org.partiql.pig.runtime.asPrimitive
+
+/**
+ * Resolves all variables by rewriting `(id <name> <case-sensitivity> <scope-qualifier>)` to
+ * `(id <name> <unique-index>)`) or `(global_id <name> <unique-id>)`.  The latter is usually a reference to
+ * a database table.  `<unique-id>` is supplied by the integrating PartiQL service by means of the [globals] specified
+ * by callers of this function.  Note that in general, all `(scan (global_id ...) ...)` operators will later be
+ * rewritten to an optimized physical read operator.
+ *
+ * The [problemHandler] is notified of any undefined variables.  Resolution does not stop on the first error, rather
+ * we keep going to provide the end user any additional error messaging, unless [ProblemHandler.handleProblem] throws
+ * an exception when an error is logged.  **If any undefined variables are detected, in order to allow traversal to
+ * continue, a fake index value is used in place of a real one and the resolved logical plan returned by this function
+ * is guaranteed to be invalid.** **Therefore, it is the responsibility therefore of callers to check if any problems
```

Comment:
A redundant _therefore`?

Reply:
Out of order, actually.  Fixing...

Reply:
also added a couple dozen lines about `$__dynamic_lookup__`